### ISSUE TITLE
Necessary update on the formal vocabulary?

### ISSUE THREAD

**Issue by iherman** [w3c/vc-data-integrity#322]
Reading through w3c/controller-document#116 it helped me to understand some things. My way of getting my thoughts in order was to try to map what I read to the [security vocabulary](https://w3c.github.io/vc-data-integrity/vocab/security/vocabulary.html) (which is, after all, simple ontology).

To check my understanding, I believe the following statements are true (some are trivial, some less):

1. _VerificationMethod_ and _ControllerDocument_ are two distinct concepts (i.e., they should be considered as distinct Classes, in RDFS parlance).
2. The _verifcationMethod_ property designates a _VerificationMethod_ instance (i.e., the _range_ of the property is a resource of type _VerificationMethod_)
3. The _VerificationMethod_ class has some subclasses defined in the controller specification, namely _Multikey_, _JsonWebKey_, and _Ed25519VerificationKey2020_.
4. The two classes in (1) have a common property, namely _controller_ (i.e., the property's _domain_ is the union of those two classes). 
5. The _ControllerDocument_ concept is ***not*** used by the Data Integrity specification, only the _VerificationMethod_ (by the way of the _verificationMethod_ property).
6. As consequence of the previous statement that it would be a mistake to define the _domain_ of the _verificationMethod_ as being a _ControllerDocument_ (i.e., ontologically one cannot restrain the classes on which it is used to be a controller document.) Actually, there should be no restraint on the domain whatsoever.

Looking at the [vocabulary](https://w3c.github.io/vc-data-integrity/vocab/security/vocabulary.html) (see also its [graphic representation](https://w3c.github.io/vc-data-integrity/vocab/security/vocabulary.svg)) we are _almost_ o.k. but not fully. The glaring (and significant) missing concept is the _ControllerDocument_. Per (1) above I believe it should be added as a separate class and, per (4) it should be an alternative domain for the _controller_ property.

(Note that the _alsoKnownAs_ and _service_ properties, though listed in the [specification](https://www.w3.org/TR/controller-document/#controller-documents) as properties on controller document, do not appear in the vocabulary or in its diagram. That is because these two properties are "borrowed" from other vocabularies.)

Long story short, I believe the following changes should be done on the vocabulary:

- Add the _ControllerDocument_ class to the vocabulary
- The _domain_ of the _controller_ property should be changed to include _ControllerDocument_ as an alternative.

I also believe that the statement (5) is not absolutely obvious from the current text, and it should be reinforced somehow...

**Comment by filip26** [comment]
add 5. 
>  Note that when verificationMethod is expressed in a [data integrity proof](https://www.w3.org/TR/vc-data-integrity/#dfn-data-integrity-proof), the value points to the actual location of the data; that is, the verificationMethod references, via a URL, the location of the [public key](https://www.w3.org/TR/vc-data-integrity/#dfn-public-key) that can be used to verify the proof. This [public key](https://www.w3.org/TR/vc-data-integrity/#dfn-public-key) data is stored in a [controller document](https://www.w3.org/TR/vc-data-integrity/#dfn-controller-document), which contains a full description of the verification method

https://www.w3.org/TR/vc-data-integrity/#proofs:~:text=Note%20that%20when,the%20verification%20method.

The paragraph language could be more straightforward, but it says that `verificationMethod`  value is an URL that dereferences to `controller-document` holding a key.

**Comment by iherman** [comment]
Thanks @filip26. But...

> add 5.
> 
> > Note that when verificationMethod is expressed in a [data integrity proof](https://www.w3.org/TR/vc-data-integrity/#dfn-data-integrity-proof), the value points to the actual location of the data; that is, the verificationMethod references, via a URL, the location of the [public key](https://www.w3.org/TR/vc-data-integrity/#dfn-public-key) that can be used to verify the proof. This [public key](https://www.w3.org/TR/vc-data-integrity/#dfn-public-key) data is stored in a [controller document](https://www.w3.org/TR/vc-data-integrity/#dfn-controller-document), which contains a full description of the verification method
> 
> https://www.w3.org/TR/vc-data-integrity/#proofs:~:text=Note%20that%20when,the%20verification%20method.
> 
> The paragraph language could be more straightforward, but it says that `verificationMethod` value is an URL that dereferences to `controller-document` holding a key.

Well, that is not the way I read that sentence! The sentence says that the:

> data is ***stored*** in a controller document" (emphasis is mine). 

My reading of this is literal: there is, somewhere, a JSON-LD document, declared as a controller document, which contains a `verificationMethod` pointing, via a URL, to a, say, `JsonWebKey` (which is a subclass of the `VerificationMethod` class). The `verificationMethod` used in a DI proof uses _**that**_ URL. There is no method defined in DI (that I see) that would point, through a URL, at the **_controller document itself_** (which would translate into a property whose range is `ControllerDocument`). Hence my statement (5) above: in a formal, vocabulary sense, the `ControllerDocument` concept is not used in DI (although I do believe that the class must be added to the vocabulary).

Yes, some clarification may indeed be needed.

cc @dlongley @msporny (my apologies not to have cc-d you before).

**Comment by filip26** [comment]
@iherman Thanks, I see your point. My interpretation is biased by how it works with DID URL when passed as a `verifificationMethod`. The DID URL is resolved to DID document/`ControllerDocument` in order to get a key.

**Comment by msporny** [comment]
@iherman if this is an update to the security vocabulary, we'll need to transfer the issue to the vc-data-integrity repo and raise the PR there. I'm ambivalent wrt. the need to do this as it won't affect any of the current implementations, AFAICT. I'm fine to do it for reasons of completeness. 

Wondering what @dlongley feels about all of the above? 

In any case, I raised PR https://github.com/w3c/vc-data-integrity/pull/320 to address this issue. This issue will be closed once PR https://github.com/w3c/vc-data-integrity/pull/320 has been merged.

**Comment by iherman** [comment]
@msporny I will take over the https://github.com/w3c/vc-data-integrity/pull/320 (see https://github.com/w3c/vc-data-integrity/pull/320#issuecomment-2481189859).

I also made a comment originally

> I also believe that the statement (5) is not absolutely obvious from the current text, and it should be reinforced somehow...

but that is probably the same discussion/issue as in https://github.com/w3c/controller-document/issues/119#issuecomment-2478149377.

**Comment by peacekeeper** [comment]
Maybe the class should really be called `Controller`, rather than `ControllerDocument`?

Does the "id" identify the controller, or the document which describes the controller? :)

**Comment by iherman** [comment]
> Maybe the class should really be called `Controller`, rather than `ControllerDocument`?
> 
> Does the "id" identify the controller, or the document which describes the controller? :)

The `id` identifies the document. The controller is identified by the `controller` property. I.e., I do not think the class name should change.

**Comment by msporny** [comment]
PR #320 has been merged, closing.
